File: backend/routes/insert.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from db.connection import get_connection

router = APIRouter()

class InsertData(BaseModel):
    sport: str
    league: str
    first: str
    second: str
    user: int

def img_search(db, variable: str, team_or_league: str) -> int:
    with db.cursor() as cursor:
        cursor.execute("SELECT id FROM images WHERE name = %s", (variable,))
        row = cursor.fetchone()
    
    img_id = row["id"] if row else None
    
    if team_or_league == "league":
        return int(img_id) if img_id is not None else 1
    else:
        return int(img_id) if img_id is not None else 2

def add_data(db, league: str, first_team: str, second_team: str, user_id: int, table_leagues: str, table_teams: str, table_broker: str):
    with db.cursor() as cursor:
        # Check and update/insert League
        cursor.execute(f"SELECT * FROM {table_leagues} WHERE name = %s AND id_user = %s", (league, user_id))
        if cursor.rowcount > 0:
            cursor.execute(f"UPDATE {table_leagues} SET count = count + 1 WHERE name = %s AND id_user = %s", (league, user_id))
        else:
            img = img_search(db, league, "league")
            cursor.execute(f"INSERT INTO {table_leagues} (id_user, name, count, img) VALUES (%s, %s, 1, %s)", (user_id, league, img))

        # Check and update/insert First Team
        cursor.execute(f"SELECT * FROM {table_teams} WHERE name = %s AND id_user = %s", (first_team, user_id))
        if cursor.rowcount > 0:
            cursor.execute(f"UPDATE {table_teams} SET home_count = home_count + 1 WHERE name = %s AND id_user = %s", (first_team, user_id))
        else:
            img = img_search(db, first_team, "team")
            cursor.execute(f"INSERT INTO {table_teams} (id_user, name, home_count, away_count, img) VALUES (%s, %s, 1, 0, %s)", (user_id, first_team, img))
            
        # Check and update/insert Second Team
        cursor.execute(f"SELECT * FROM {table_teams} WHERE name = %s AND id_user = %s", (second_team, user_id))
        if cursor.rowcount > 0:
            cursor.execute(f"UPDATE {table_teams} SET away_count = away_count + 1 WHERE name = %s AND id_user = %s", (second_team, user_id))
        else:
            img = img_search(db, second_team, "team")
            cursor.execute(f"INSERT INTO {table_teams} (id_user, name, home_count, away_count, img) VALUES (%s, %s, 0, 1, %s)", (user_id, second_team, img))

        # Get League ID
        cursor.execute(f"SELECT id FROM {table_leagues} WHERE name = %s AND id_user = %s", (league, user_id))
        league_row = cursor.fetchone()
        if league_row:
            league_id = league_row["id"]
        else:
            # Fallback grab
            cursor.execute(f"SELECT id FROM {table_leagues} WHERE name = %s", (league,))
            league_row_fallback = cursor.fetchone()
            league_id = league_row_fallback["id"] if league_row_fallback else 1

        teams = [first_team, second_team]
        team_ids = []
        for team in teams:
            cursor.execute(f"SELECT id FROM {table_teams} WHERE name = %s AND id_user = %s", (team, user_id))
            team_row = cursor.fetchone()
            if team_row:
                team_ids.append(team_row["id"])
            else:
                cursor.execute(f"SELECT id FROM {table_teams} WHERE name = %s", (team,))
                fallback_row = cursor.fetchone()
                if fallback_row:
                    team_ids.append(fallback_row["id"])

        for team_id in team_ids:
            cursor.execute(f"INSERT IGNORE INTO {table_broker} (id_league, id_team) VALUES (%s, %s)", (league_id, team_id))

    db.commit()


@router.post("/insert/insert-data")
def insert_data(data: InsertData):
    if not data.sport or not data.league or not data.first or not data.second or not data.user:
        raise HTTPException(status_code=400, detail="Invalid data")

    db = get_connection()
    try:
        if data.sport == "League of Legends":
            add_data(db, data.league, data.first, data.second, data.user, "lol_leagues", "lol_teams", "lol_broker")
        elif data.sport == "Counter Strike":
            add_data(db, data.league, data.first, data.second, data.user, "cs_leagues", "cs_teams", "cs_broker")
        else:
            sport_prefix = data.sport.lower()
            season_league = f"{sport_prefix}_league_season"
            season_team = f"{sport_prefix}_teams_season"
            season_broker = f"{sport_prefix}_broker_season"
            
            year_league = f"{sport_prefix}_league_year"
            year_team = f"{sport_prefix}_teams_year"
            year_broker = f"{sport_prefix}_broker_year"
            
            add_data(db, data.league, data.first, data.second, data.user, season_league, season_team, season_broker)
            add_data(db, data.league, data.first, data.second, data.user, year_league, year_team, year_broker)
        return {"success": True}
    finally:
        db.close()

File: website/api/insert.php
<?php

class Insert
{
    private $db;

    public function __construct($conn)
    {
        $this->db = $conn;
    }

    public function insertData()
    {
        $data = json_decode(file_get_contents("php://input"), true);
        $sport = $data['sport'];
        $league = $data['league'];
        $firstTeam = $data['first'];
        $secondTeam = $data['second'];
        $userID = $data['user'];

        if (!$sport || !$league || !$firstTeam || !$secondTeam || !$userID) {
            http_response_code(400);
            echo json_encode(["error" => "Invalid data"]);
            exit;
        }

        switch ($sport) {
            case "League of Legends":
                $tabelLeagues = "lol_leagues";
                $tabelTeams = "lol_teams";
                $tabelBroker = "lol_broker";

                $this->addData($league, $firstTeam, $secondTeam, $userID, $tabelLeagues, $tabelTeams, $tabelBroker);
                break;
            case "Counter Strike":
                $tabelLeagues = "cs_leagues";
                $tabelTeams = "cs_teams";
                $tabelBroker = "cs_broker";

                $this->addData($league, $firstTeam, $secondTeam, $userID, $tabelLeagues, $tabelTeams, $tabelBroker);
                break;
            default:
                $seasonLeague = strtolower($sport) . "_leagues_season";
                $seasonTeam = strtolower($sport) . "_teams_season";
                $seasonBroker = strtolower($sport) . "_broker_season";

                $yearLeague = strtolower($sport) . "_leagues_year";
                $yearTeam = strtolower($sport) . "_teams_year";
                $yearBroker = strtolower($sport) . "_broker_year";

                $this->addData($league, $firstTeam, $secondTeam, $userID, $seasonLeague, $seasonTeam, $seasonBroker);
                $this->addData($league, $firstTeam, $secondTeam, $userID, $yearLeague, $yearTeam, $yearBroker);
        }
    }

    private function addData(string $league, string $firstTeam, string $secondTeam, string $userID, string $tabelLeagues, string $tabelTeams, string $tabelBroker)
    {
        $this->db->begin_transaction();
        $sqlLeagueCheck = "SELECT * FROM $tabelLeagues WHERE name = ? AND id_user = ?";
        $sqlLeagueCheck = $this->db->prepare($sqlLeagueCheck);
        $sqlLeagueCheck->bind_param("ss", $league, $userID);
        $sqlLeagueCheck->execute();
        $result = $sqlLeagueCheck->get_result();

        if ($result->num_rows > 0) {
            $sqlUpdateLeague = "UPDATE $tabelLeagues SET count = count + 1 WHERE name = ? AND id_user = ?";
            $query = $this->db->prepare($sqlUpdateLeague);
            $query->bind_param("ss", $league, $userID);
            $query->execute();
        }
        else {
            $img = $this->imgSearch($league, "league");

            $sqlInsertLeague = "INSERT INTO $tabelLeagues (id, id_user, name, count, img) VALUES (NULL, ?, ?, 1, ?)";
            $query = $this->db->prepare($sqlInsertLeague);
            $query->bind_param("isi", $userID, $league, $img);
            $query->execute();
        }

        $sqlFirstTeamCheck = "SELECT * FROM $tabelTeams WHERE name = ? AND id_user = ?";
        $sqlFirstTeamCheck = $this->db->prepare($sqlFirstTeamCheck);
        $sqlFirstTeamCheck->bind_param("ss", $firstTeam, $userID);
        $sqlFirstTeamCheck->execute();
        $result = $sqlFirstTeamCheck->get_result();

        if ($result->num_rows > 0) {
            $sqlUpdateFirst = "UPDATE $tabelTeams SET home_count = home_count + 1 WHERE name = ? AND id_user = ?";
            $query = $this->db->prepare($sqlUpdateFirst);
            $query->bind_param("ss", $league, $userID);
            $query->execute();
        }
        else {
            $img = $this->imgSearch($firstTeam, "team");

            $sqlFirstTeamInsert = "INSERT INTO $tabelTeams (id, id_user, name, home_count, away_count, img) VALUES (NULL, ?, ?, 1, 0, ?)";
            $query = $this->db->prepare($sqlFirstTeamInsert);
            $query->bind_param("isi", $userID, $league, $img);
            $query->execute();
        }

        $sqlSecondTeamCheck = "SELECT * FROM $tabelTeams WHERE name = ? AND id_user = ?";
        $sqlSecondTeamCheck = $this->db->prepare($sqlSecondTeamCheck);
        $sqlSecondTeamCheck->bind_param("ss", $secondTeam, $userID);
        $sqlSecondTeamCheck->execute();
        $result = $sqlSecondTeamCheck->get_result();

        if ($result->num_rows > 0) {
            $sqlUpdateSecond = "UPDATE $tabelTeams SET away_count = away_count + 1 WHERE name = ? AND id_user = ?";
            $query = $this->db->prepare($sqlUpdateSecond);
            $query->bind_param("ss", $league, $userID);
            $query->execute();
        }
        else {
            $img = $this->imgSearch($secondTeam, "team");

            $sqlSecondTeamInsert = "INSERT INTO $tabelTeams (id, id_user, name, home_count, away_count, img) VALUES (NULL, ?, ?, 0, 1, ?)";
            $query = $this->db->prepare($sqlSecondTeamInsert);
            $query->bind_param("isi", $userID, $league, $img);
            $query->execute();
        }

        $sqlTakeIDLeague = "SELECT id FROM $tabelLeagues WHERE name = ?";
        $sqlTakeIDLeague = $this->db->prepare($sqlTakeIDLeague);
        $sqlTakeIDLeague->bind_param("s", $league);
        $sqlTakeIDLeague->execute();
        $LeagueID = $sqlTakeIDLeague->get_result()->fetch_assoc()['id'];

        $teams = [$firstTeam, $secondTeam];
        $teamsIDs = [];
        foreach ($teams as $team) {
            $sqlTakeIDTeam = "SELECT id FROM $tabelTeams WHERE name = ?";
            $sqlTakeIDTeam = $this->db->prepare($sqlTakeIDTeam);
            $sqlTakeIDTeam->bind_param("s", $team);
            $sqlTakeIDTeam->execute();
            $result = $sqlTakeIDTeam->get_result()->fetch_assoc()['id'];

            $teamsIDs[] = $result;
        }

        foreach ($teamsIDs as $team) {
            $sqlInsertBroker = "INSERT IGNORE INTO $tabelBroker VALUES (?, ?)";
            $query = $this->db->prepare($sqlInsertBroker);
            $query->bind_param("ii", $LeagueID, $team);
            $query->execute();
        }
        $this->db->commit();
    }

    private function imgSearch($varible, $teamOrLeague)
    {
        $sqlIMGSearch = "SELECT id FROM images WHERE name = ?";
        $sqlIMGSearch = $this->db->prepare($sqlIMGSearch);
        $sqlIMGSearch->bind_param("s", $varible);
        $sqlIMGSearch->execute();
        $result = $sqlIMGSearch->get_result()->fetch_assoc();
        $imgID = $result ? $result['id'] : null;

        if ($teamOrLeague == "league") {
            if ($imgID === null) {
                $img = 1;
            }
            else {
                $img = $imgID;
            }
        }
        else {
            if ($imgID === null) {
                $img = 2;
            }
            else {
                $img = $imgID;
            }
        }
        return $img;
    }
}
